Islandora-CLAW/islandora
# Using Context instead of Rules

**GitHub Issue**: https://github.com/Islandora-CLAW/CLAW/issues/765

Other Relevant Links:
* https://github.com/Islandora-CLAW/CLAW/issues/656
* https://github.com/Islandora-CLAW/CLAW/issues/682 

# What does this Pull Request do?

Replaces Rules with Context

# What's new?

* Three contexts (file, media, node) which have execute preconfigued actions on write events.
* Preconfigured actions for emitting events to queues (indexing in triplestore and fedora)
* 'Views bulk operation' functionality since everything is actions (essentially https://github.com/Islandora/islandora_bookmark functionality is baked into core) 
* Custom Condition plugins for file, media, and node content types.
* Functional tests! 

# How should this be tested?

Using https://github.com/islandora-devops/claw-playbook
- `vagrant up`
- `vagrant ssh`

Then uninstall the module, which is a bit of a pain right now (see https://github.com/Islandora-CLAW/CLAW/issues/685)
- `/var/www/html/drupal/vendor/bin/drupal config:delete active views.view.file_checksum` (This PR fixes the bug that leaves the view in the database)
- `/var/www/html/drupal/vendor/bin/drupal config:delete active field.field.media.tn.field_height`
- `/var/www/html/drupal/vendor/bin/drupal config:delete active field.field.media.tn.field_width`
- `/var/www/html/drupal/vendor/bin/drupal config:delete active field.field.media.tn.field_mimetype`
- `/var/www/html/drupal/vendor/bin/drupal config:delete active field.field.media.tn.field_image`
- `/var/www/html/drupal/vendor/bin/drush pmu islandora`

Pull in dependencies with composer
- `cd /var/www/html/drupal`
- `composer require drupal/context`

Then pull in the changes for this PR
- `cd /var/www/html/drupal/web/modules/contrib/islandora`
- `git remote add danny https://github.com/dannylamb/islandora.git`
- `git fetch danny`
- `git checkout context-preconfigured`

Then install
- `/var/www/html/drupal/vendor/bin/drush en islandora_image context_ui`

Check that fedora and triplestore indexing still function as expected.

Then go check out `admin/config/system/actions` and inspect all the preconfigured actions for publishing events to queues.
![screenshot from 2017-12-19 16-59-41](https://user-images.githubusercontent.com/20773151/34178605-2f5fa452-e4de-11e7-90cd-a218d6bf200c.png)


You can make a preconfigured action to print something to the screen:
![screenshot from 2017-12-19 17-03-12](https://user-images.githubusercontent.com/20773151/34178734-9f95c36e-e4de-11e7-86fe-e5d2792d0a32.png)

 
![screenshot from 2017-12-19 17-04-56](https://user-images.githubusercontent.com/20773151/34178787-c5390a68-e4de-11e7-8897-41c9bcf90109.png)


Then you can execute that action as a Context reaction adding it as an `index` or `deindex` action (I'm open to better naming for those).  For example, to print the message every time a node is updated or created, go to `admin/structure/context/node` and Ctrl+Click the `Say Hey` action in the `Index` select box.
![screenshot from 2017-12-19 17-09-22](https://user-images.githubusercontent.com/20773151/34178954-6bd5e846-e4df-11e7-96cd-f6064535d346.png)


FYI, the usual Shift+Click and Ctrl+Click functionality is available when using that select box.


Click `Save and continue`.  Then go create or update a node and you should see this:
![screenshot from 2017-12-19 17-11-18](https://user-images.githubusercontent.com/20773151/34179026-b0243f5c-e4df-11e7-9816-95cbadce5693.png)


You can also do cool bookmark type stuff since everything is core actions.  Check out the main admin view for nodes to see all the things you can do.  
![screenshot from 2017-12-19 17-13-40](https://user-images.githubusercontent.com/20773151/34179249-84d8e842-e4e0-11e7-8e8c-661e5706122b.png)



There's also stuff available for Media.
![screenshot from 2017-12-19 17-15-11](https://user-images.githubusercontent.com/20773151/34179154-37280970-e4e0-11e7-9b42-d9cd70ec04ad.png)


# Interested parties
@Islandora-CLAW/committers @rosiel @mjordan
